### PR TITLE
[MISC] Store layout in a struct instead of writing directly to stream.

### PR DESCRIPTION
**Pull request by smehringer**
Big step towards having an HIBF data structure that does layouting and building without storing the layout in between.

### CODE REVIEW

File: include/chopper/layout/layout.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+// Currently, the layout is structured by user bin.
+struct layout
+{
+    struct max_bin
+    {
+        std::vector<size_t> previous_TB_indices{}; // identifies the IBF based on upper levels
+        size_t id{};                               // the technical bin id that has the maximum kmer content
+
+        auto operator<=>(max_bin const &) const = default;
```
Comment (eseiler) [inline include/chopper/layout/layout.hpp:20]: Note: add suggestion

File: include/chopper/layout/layout.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    };
+
+    struct user_bin
+    {
+        std::string id{};                          // the id of the user bin. Often its filename.
+        std::vector<size_t> previous_TB_indices{}; // previous technical bin indices which refer to merged bin indices.
+        size_t number_of_technical_bins{};         // 1 == signle bin, >1 == split_bin
+        size_t storage_TB_id{}; // the id of the technical bin that the user bin is actullly stored in
+
+        auto operator<=>(user_bin const &) const = default;
```
Comment (eseiler) [inline include/chopper/layout/layout.hpp:46]: Note: add suggestion

File: include/chopper/layout/layout.hpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+        // needs a template (instead of using std::ostream directly) to be able to only include <iosfwd>
+        template <typename stream_type>
```
Comment (eseiler) [inline include/chopper/layout/layout.hpp:49]:
```cpp
template <typename stream_type = std::basic_ostream<char>>
        requires requires (stream_type stream, size_t value) { stream << value; }
    friend stream_type & operator<<
```

File: include/chopper/layout/layout.hpp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+{
+    struct max_bin
+    {
+        std::vector<size_t> previous_TB_indices{}; // identifies the IBF based on upper levels
+        size_t id{};                               // the technical bin id that has the maximum kmer content
+
+        auto operator<=>(max_bin const &) const = default;
+
+        // needs a template (instead of using std::ostream directly) to be able to only include <iosfwd>
+        template <typename stream_type>
```
Comment (eseiler) [inline include/chopper/layout/layout.hpp:23]:
```suggestion
        template <typename stream_type = std::ostream>
```

File: include/chopper/layout/layout.hpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            requires std::derived_from<stream_type, std::ostream>
+        friend stream_type & operator<<(stream_type & stream, max_bin const & object)
+        {
+            stream << prefix::header << prefix::merged_bin << "_";
```
Comment (eseiler) [inline include/chopper/layout/layout.hpp:27]:
```suggestion
            stream << prefix::header << prefix::merged_bin << '_';
```

File: include/chopper/layout/layout.hpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            auto it = object.previous_TB_indices.begin();
+            if (it != object.previous_TB_indices.end())
+                stream << *it++;
+            for (; it != object.previous_TB_indices.end(); ++it)
+                stream << ';' << *it;
```
Comment (eseiler) [inline include/chopper/layout/layout.hpp:32]:
```suggestion
            auto it = object.previous_TB_indices.begin();
            auto end = object.previous_TB_indices.end();
            // If not empty, we join with ';'
            if (it != end)
            {
                stream << *it;
                while (++it != end)
                    stream << ';' << *it;
            }
```

File: include/chopper/layout/layout.hpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        template <typename stream_type>
+            requires std::derived_from<stream_type, std::ostream>
+        friend stream_type & operator<<(stream_type & stream, user_bin const & object)
+        {
+            stream << object.id << '\t';
+            for (auto bin : object.previous_TB_indices)
+                stream << bin << ';';
+            stream << object.storage_TB_id << '\t';
+            for (size_t i = 0; i < object.previous_TB_indices.size(); ++i) // number of bins per merged level is 1
+                stream << '1' << ';';
```
Comment (eseiler) [inline include/chopper/layout/layout.hpp:58]:
```suggestion
            for ([[maybe_unused]] auto && elem : object.previous_TB_indices) // number of bins per merged level is 1
                stream << "1;";
```

I think it's easier to read, it's more clear that `elem` (`i`) is unused. `i` and `1` also look rather similar, so it's easy to presume a typo.

File: include/chopper/layout/output.hpp [hunk 8]
```diff
@@ -24,14 +25,18 @@ inline void write_config_to(configuration const & config, std::ostream & stream)
...
-    write_config_to(config, stream);
     stream << prefix::header << prefix::high_level << " max_bin_id:" << max_hibf_id << '\n';
-    stream << header;
+    for (auto const & max_bin : hibf_layout.max_bins)
+        stream << max_bin << '\n';
+}
+
+inline void write_layout_content_to(layout const & hibf_layout, std::ostream & stream)
+{
+    stream << prefix::header << "FILES\tBIN_INDICES\tNUMBER_OF_BINS" << std::endl;
```
Comment (eseiler) [inline include/chopper/layout/output.hpp:37]:
```suggestion
    stream << prefix::header << "FILES\tBIN_INDICES\tNUMBER_OF_BINS\n";
```

File: include/chopper/layout/print_result_line.hpp [hunk 9]
```diff
@@ -3,40 +3,22 @@
...
-                        << '\t' << data.previous.estimated_sizes << (is_top_level ? "" : ";") << average_bin_size
-                        << '\t' << data.previous.optimal_score << (is_top_level ? "" : ";") << optimal_score << '\t'
-                        << data.previous.correction << (is_top_level ? "" : ";") << correction << '\t'
-                        << data.previous.tmax << (is_top_level ? "" : ";") << num_technical_bins << '\n';
+    stream << user_bin.id << '\t';
+    for (auto bin : user_bin.previous_TB_indices)
+        stream << bin << ';';
+    stream << user_bin.storage_TB_id << '\t';
+    for (size_t i = 0; i < user_bin.previous_TB_indices.size(); ++i) // number of bins per merged level is 1
+        stream << '1' << ';';
```
Comment (eseiler) [inline include/chopper/layout/print_result_line.hpp:19]:
```suggestion
    for ([[maybe_unused]] auto && elem : user_bin.previous_TB_indices) // number of bins per merged level is 1
        stream << "1;";
```

File: include/chopper/layout/print_result_line.hpp [hunk 10]
```diff
@@ -3,40 +3,22 @@
...
+    stream << user_bin.number_of_technical_bins;
+    stream << '\n';
```
Comment (eseiler) [inline include/chopper/layout/print_result_line.hpp:21]:
```suggestion
    stream << user_bin.number_of_technical_bins << '\n';
```

File: include/chopper/layout/layout.hpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        std::string id{};                          // the id of the user bin. Often its filename.
+        std::vector<size_t> previous_TB_indices{}; // previous technical bin indices which refer to merged bin indices.
+        size_t number_of_technical_bins{};         // 1 == signle bin, >1 == split_bin
+        size_t storage_TB_id{}; // the id of the technical bin that the user bin is actullly stored in
+
+        auto operator<=>(user_bin const &) const = default;
+
+        // needs a template (instead of using std::ostream directly) to be able to only include <iosfwd>
+        template <typename stream_type>
```
Comment (eseiler) [inline include/chopper/layout/layout.hpp:49]:
```suggestion
        template <typename stream_type = std::ostream>
```

Comment (smehringer) [reply include/chopper/layout/layout.hpp:53]: This causes an ambiguous overload error with the debug_stream_type...

<details>
```bash
/home/mi/smehringer/AGABI/Chopper/lib/seqan3/include/seqan3/core/debug_stream/range.hpp:100:27: error: static assertion failed: The reference type of the passed range cannot be streamed into the debug_stream.
  100 |     static_assert(detail::reference_type_is_streamable_v<rng_t, char_t>,
      |                   ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/mi/smehringer/AGABI/Chopper/lib/seqan3/include/seqan3/core/debug_stream/range.hpp:100:27: note: ‘seqan3::detail::reference_type_is_streamable_v<const std::vector<chopper::layout::layout::max_bin>&, char>’ evaluates to false
/home/mi/smehringer/AGABI/Chopper/lib/seqan3/include/seqan3/core/debug_stream/range.hpp:108:11: error: ambiguous overload for ‘operator<<’ (operand types are ‘seqan3::debug_stream_type<char>’ and ‘const chopper::layout::layout::max_bin’)
  108 |         s << *b;
      |         ~~^~~~~

```
</details>

**Pull request merged**